Source/WebCore/platform/gstreamer/GStreamerElementHarness.cpp

```diff
@@ -0,0 +1,354 @@
+    }
+
+    GstSegment segment;
+    gst_segment_init(&segment, GST_FORMAT_TIME);
+    pushEvent(gst_event_new_segment(&segment));
+
+    m_stickyEventsSent.store(true);
+}
+
+bool GStreamerElementHarness::pushSample(GstSample* sample)
+{
+    GRefPtr<GstCaps> caps = gst_sample_get_caps(sample);
+    GST_TRACE_OBJECT(m_element.get(), "Pushing sample with caps %" GST_PTR_FORMAT, caps.get());
+    if (!m_playing.load())
+        start(WTFMove(caps));
+    else {
+        auto currentCaps = adoptGRef(gst_pad_get_current_caps(m_srcPad.get()));
+        GST_TRACE_OBJECT(m_element.get(), "Current caps: %" GST_PTR_FORMAT, currentCaps.get());
+        if (!currentCaps || gst_pad_needs_reconfigure(m_srcPad.get()))
+            pushStickyEvents(WTFMove(caps));
+    }
+    return pushBuffer(gst_buffer_ref(gst_sample_get_buffer(sample)));
+}
+
+bool GStreamerElementHarness::pushBuffer(GstBuffer* buffer)
+{
+    if (!m_stickyEventsSent.load())
+        return false;
+
+    GST_TRACE_OBJECT(m_element.get(), "Pushing %" GST_PTR_FORMAT, buffer);
+    auto result = gst_pad_push(m_srcPad.get(), buffer);
+    GST_TRACE_OBJECT(m_element.get(), "Buffer push result: %s", gst_flow_get_name(result));
+    return result == GST_FLOW_OK;
+}
+
+bool GStreamerElementHarness::pushEvent(GstEvent* event)
+{
+    GST_TRACE_OBJECT(m_element.get(), "Pushing %" GST_PTR_FORMAT, event);
+    auto result = gst_pad_push_event(m_srcPad.get(), event);
+    GST_TRACE_OBJECT(m_element.get(), "Result: %s", boolForPrinting(result));
+    return result;
+}
+
+GRefPtr<GstBuffer> GStreamerElementHarness::pullBuffer()
+{
+    GST_LOG_OBJECT(m_element.get(), "%zu buffers currently queued", m_bufferQueue.size());
+    Locker locker { m_bufferQueueLock };
+    if (m_bufferQueue.isEmpty())
+        return nullptr;
+    return m_bufferQueue.takeLast();
+}
+
+GRefPtr<GstEvent> GStreamerElementHarness::pullEvent()
+{
+    Locker locker { m_sinkEventQueueLock };
+    if (m_sinkEventQueue.isEmpty())
+        return nullptr;
+    return m_sinkEventQueue.takeLast();
+}
+
+const GRefPtr<GstCaps>& GStreamerElementHarness::outputCaps()
+{
+    if (m_outputCaps)
+        return m_outputCaps;
+
+    auto pad = adoptGRef(gst_element_get_static_pad(m_element.get(), "src"));
+    m_outputCaps = adoptGRef(gst_pad_get_current_caps(pad.get()));
+    GST_DEBUG_OBJECT(m_element.get(), "Output caps: %" GST_PTR_FORMAT, m_outputCaps.get());
+    return m_outputCaps;
+}
+
+GstFlowReturn GStreamerElementHarness::chainBuffer(GstBuffer* outputBuffer)
+{
+    Locker locker { m_bufferQueueLock };
+    auto buffer = adoptGRef(outputBuffer);
+    m_bufferQueue.prepend(WTFMove(buffer));
+    return GST_FLOW_OK;
+}
+
+bool GStreamerElementHarness::sinkQuery(GstPad* pad, GstObject* parent, GstQuery* query)
+{
+    bool result = TRUE;
+    switch (GST_QUERY_TYPE(query)) {
+    case GST_QUERY_CAPS: {
+        GstCaps* filter = nullptr;
+        GRefPtr<GstCaps> caps;
+        if (m_outputCaps)
+            caps = m_outputCaps.get();
```

Comment:
Why do you need to keep the `.get()` ? I think you don't

Reply:
Actually, the way it was before, I think it was leaking.

Reply:
Good catch!